# Data Understanding

## Objective

The purpose of this section is to understand the structure and characteristics of the raw dataset before applying any cleaning or preprocessing steps.

In [1]:
import pandas as pd
from pathlib import Path

In [2]:
df = pd.read_csv(Path('../data/raw/works_with_composers.csv'))

In [3]:
df.head()

,title,subtitle,searchterms,popular,recommended,work_id,genre,composer_id,name,complete_name,birth,death,epoch,portrait
0,China Gates,NaN,NaN,1,1,16893,Keyboard,149,Adams,John Adams,1947-01-01,NaN,21st Century,https://assets.openopus.org/portraits/74462091...
1,Doctor Atomic Symphony,NaN,NaN,0,1,16878,Orchestral,149,Adams,John Adams,1947-01-01,NaN,21st Century,https://assets.openopus.org/portraits/74462091...
2,Harmonielehre,NaN,NaN,1,1,16875,Orchestral,149,Adams,John Adams,1947-01-01,NaN,21st Century,https://assets.openopus.org/portraits/74462091...
3,Nixon in China,Opera,NaN,1,1,16890,Stage,149,Adams,John Adams,1947-01-01,NaN,21st Century,https://assets.openopus.org/portraits/74462091...
4,Short Ride in a Fast Machine,NaN,NaN,1,1,16896,Orchestral,149,Adams,John Adams,1947-01-01,NaN,21st Century,https://assets.openopus.org/portraits/74462091...


In [4]:
df.shape

(13014, 14)

In [5]:
df.columns

Index(['title', 'subtitle', 'searchterms', 'popular', 'recommended', 'work_id',
       'genre', 'composer_id', 'name', 'complete_name', 'birth', 'death',
       'epoch', 'portrait'],
      dtype='str')

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 13014 entries, 0 to 13013
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   title          13014 non-null  str  
 1   subtitle       1288 non-null   str  
 2   searchterms    57 non-null     str  
 3   popular        13014 non-null  int64
 4   recommended    13014 non-null  int64
 5   work_id        13014 non-null  int64
 6   genre          13014 non-null  str  
 7   composer_id    13014 non-null  int64
 8   name           13014 non-null  str  
 9   complete_name  13014 non-null  str  
 10  birth          13014 non-null  str  
 11  death          12747 non-null  str  
 12  epoch          13014 non-null  str  
 13  portrait       13014 non-null  str  
dtypes: int64(4), str(10)
memory usage: 1.4 MB


In [7]:
df.describe()

,popular,recommended,work_id,composer_id
count,13014.000000,13014.000000,13014.000000,13014.00000
mean,0.021592,0.059244,14839.250807,127.75972
std,0.145353,0.236090,8105.524462,63.27243
min,0.000000,0.000000,24.000000,2.00000
25%,0.000000,0.000000,7936.250000,82.00000
50%,0.000000,0.000000,15158.500000,129.00000
75%,0.000000,0.000000,22149.750000,188.00000
max,1.000000,1.000000,28122.000000,220.00000


In [8]:
df.describe(include=str)

,title,subtitle,searchterms,genre,name,complete_name,birth,death,epoch,portrait
count,13014,1288,57,13014,13014,13014,13014,12747,13014,13014
unique,12887,219,57,5,77,77,68,69,10,77
top,Violin Concerto,Opera,"requiem, grande messe des morts",Vocal,Bach,Johann Sebastian Bach,1685-01-01,1750-01-01,Baroque,https://assets.openopus.org/portraits/12091447...
freq,6,335,1,5252,981,981,1929,981,3005,981


In [9]:
df.duplicated().sum()

np.int64(0)

In [10]:
df['work_id'].duplicated().sum()

np.int64(0)

In [11]:
df.value_counts('title')

title
Violin Concerto                      6
Symphony no. 3                       5
Khovanshchina                        5
4 Songs, op. 2                       4
Piano Concerto                       4
                                    ..
Slow Movement, for string quartet    1
Sonatasatz                           1
String Trio, op. 20                  1
String Trio movement                 1
Variations, op. 30                   1
Name: count, Length: 12887, dtype: int64

In [12]:
df.value_counts('subtitle')

subtitle
Opera                               335
Song                                204
Songs                               105
Ballet                               64
Song cycle                           51
                                   ... 
For tenor and orchestra               1
Four violin concertos, op. 8          1
For military band                     1
Version completed by Felix Mottl      1
For band                              1
Name: count, Length: 219, dtype: int64

In [13]:
df.value_counts('searchterms')

searchterms
requiem, grande messe des morts                                                                 1
Sonata for violin and piano                                                                     1
m                                                                                               1
El sombrero de tres picos, The three-cornered hat                                               1
Le Chasseur maudit                                                                              1
Symphony in D minor                                                                             1
Symphonic variations,Variations symphoniques                                                    1
second rhapsody, rhapsody no. 2                                                                 1
2 pages, two pages                                                                              1
Po zarostlém chodnícku                                                                          1
sinfonie

In [14]:
df.value_counts('genre')

genre
Vocal         5252
Keyboard      3069
Orchestral    2117
Chamber       1903
Stage          673
Name: count, dtype: int64

In [15]:
df.value_counts('name')

name
Bach         981
Schubert     862
Haydn        618
Vivaldi      597
Scarlatti    588
            ... 
Boulez        27
Reich         26
Mahler        24
Léonin        12
Pérotin        9
Name: count, Length: 77, dtype: int64

In [16]:
df.value_counts('complete_name')

complete_name
Johann Sebastian Bach    981
Franz Schubert           862
Franz Joseph Haydn       618
Antonio Vivaldi          597
Domenico Scarlatti       588
                        ... 
Pierre Boulez             27
Steve Reich               26
Gustav Mahler             24
Léonin                    12
Pérotin                    9
Name: count, Length: 77, dtype: int64

In [17]:
df.value_counts('birth')

birth
1685-01-01    1929
1797-01-01     862
1732-01-01     618
1678-01-01     597
1756-01-01     557
              ... 
1925-01-01      27
1936-01-01      26
1860-01-01      24
1135-01-01      12
1160-01-01       9
Name: count, Length: 68, dtype: int64

In [18]:
df.value_counts('epoch')

epoch
Baroque           3005
Romantic          2309
20th Century      2214
Early Romantic    1364
Classical         1205
Renaissance       1175
Late Romantic     1139
Post-War           479
Medieval            97
21st Century        27
Name: count, dtype: int64

## Data Understanding Summary

In this section, the structure and characteristics of the raw classical music dataset were investigated.

The dataset contains 13014 musical works and 14 features collected from the Open Opus API, including title, genre, composer's name, birth and death dates and musical era.

Initial inspection showed that:
- No duplicate records were found.
- Some features contain missing values.
- Musical form is not directly available and can be extracted from titles.
- Date columns need transformation into year format.
- Some redundant and unnecessary columns need removal.

# Data Cleaning

## Objective

The goal of this notebook is to clean and prepare raw dataset collected from the Open Opus API.

The following steps will be performed:
- Remove features that do not provide useful information for the recommendation system.
- Remove redundant composer name information.
- Convert composer birth and death dates from full date format into year format.

In [19]:
df_clean = df.drop(['portrait', 'searchterms', 'name'], axis=1)

In [20]:
df_clean = df_clean.rename(columns={'complete_name': 'composer_name'})

In [21]:
df_clean.columns

Index(['title', 'subtitle', 'popular', 'recommended', 'work_id', 'genre',
       'composer_id', 'composer_name', 'birth', 'death', 'epoch'],
      dtype='str')

In [22]:
df_clean['birth'] = pd.to_datetime(df_clean['birth']).dt.year
df_clean['death'] = pd.to_datetime(df_clean['death']).dt.year

In [23]:
df_clean[['birth', 'death']].head()

,birth,death
0,1947,NaN
1,1947,NaN
2,1947,NaN
3,1947,NaN
4,1947,NaN


In [24]:
df_clean.isna().sum()

title                0
subtitle         11726
popular              0
recommended          0
work_id              0
genre                0
composer_id          0
composer_name        0
birth                0
death              267
epoch                0
dtype: int64

### Save Cleaned Dataset

In [25]:
df_clean.to_csv(Path('../data/processed/works_with_composers_cleaned.csv'), index=False)

## Data Cleaning Summary

In this section, the raw dataset was cleaned and prepared for the next stages of the project.

The following changes were applied:
- Removed unnecessary features such as portrait, search terms and redundant composer name information.
- Transformed composer birth and death dates from full date format into year format.
- Noticed that death year is NaN for some composers. It might mean they are still alive. They will stay NaN for further analysis.
- The subtitle column also has several null values, but it may be useful for feature engineering to extract new features from the combination of title and subtitle.